### PR TITLE
Interpreter branch opcodes

### PR DESCRIPTION
**Pull request by ratik21**
- Added opcodes: `b`, `bnz`, `bz`, `Return`
- Added tests

### CODE REVIEW

File: packages/algorand-js/src/interpreter/interpreter.ts [hunk 1]
```diff
@@ -18,6 +19,8 @@ export class Interpreter {
   tx: Txn;
   gtxs: Txn[];
   accounts: AccountsMap;
+  instructions: Operator[];
+  i: number;
```
Comment (amityadav0) [inline packages/algorand-js/src/interpreter/interpreter.ts:23]: can we name it something else ? so that it is more clear why we are using this variable

Comment (ratik21) [reply packages/algorand-js/src/interpreter/interpreter.ts:23]: Can you suggest something? how about `instructionIndex`?

Comment (robert-zaremba) [reply packages/algorand-js/src/interpreter/interpreter.ts:23]: BTW, in the future we will need to rework it - when reporting error, we need to show a line in the source file, not the instruction index.

Comment (ratik21) [reply packages/algorand-js/src/interpreter/interpreter.ts:23]: yes, good point.

File: packages/algorand-js/src/errors/errors-list.ts [hunk 2]
```diff
@@ -133,6 +133,18 @@ by an index that does not exist.`
...
+    number: 19,
+    message: "Label not found",
+    title: "Label %label% not found",
+    description: `Label %label% not found`
+  },
+  INVALID_STACK_ELEM: {
+    number: 20,
+    message: "Invalid top of stack or length of stack > 1",
+    title: "Invalid top of stack or length of stack > 1",
+    description: `Either Length of stack is >= 1 or top of stack is not uint64`
```
Comment (amityadav0) [inline packages/algorand-js/src/errors/errors-list.ts:147]: Please make sure whether length of stack is `>` or `>=` 1.

File: packages/algorand-js/src/interpreter/opcode-list.ts [hunk 3]
```diff
@@ -770,6 +770,126 @@ export class Gtxna extends Op {
...
+  readonly label: string;
+
+    /**
+   * @param args words list extracted from line
+   * @param line line number in TEAL file
+   */
+  constructor (args: string[], line: number) {
+    super();
+    assertLen(args.length, 1, line);
+    this.label = args[0];
```
Comment (amityadav0) [inline packages/algorand-js/src/interpreter/opcode-list.ts:784]: `args` received will be `label:`. store it in `this.label` accordingly

File: packages/algorand-js/src/interpreter/opcode-list.ts [hunk 4]
```diff
@@ -770,6 +770,126 @@ export class Gtxna extends Op {
...
+
+// branch unconditionally to label
+export class Branch extends Op {
+  readonly label: string;
+  readonly interpreter: Interpreter;
+
+  /**
+   * @param args words list extracted from line
+   * @param line line number in TEAL file
+   * @param interpreter interpreter
```
Comment (amityadav0) [inline packages/algorand-js/src/interpreter/opcode-list.ts:798]:
```suggestion
   * @param interpreter object
```

File: packages/algorand-js/src/interpreter/opcode-list.ts [hunk 5]
```diff
@@ -770,6 +770,126 @@ export class Gtxna extends Op {
...
+
+// branch conditionally if top of stack is zero
+export class BranchIfZero extends Op {
+  readonly label: string;
+  readonly interpreter: Interpreter;
+
+  /**
+   * @param args words list extracted from line
+   * @param line line number in TEAL file
+   * @param interpreter interpreter
```
Comment (amityadav0) [inline packages/algorand-js/src/interpreter/opcode-list.ts:820]:
```suggestion
   * @param interpreter interpreter object
```

File: packages/algorand-js/src/interpreter/opcode-list.ts [hunk 6]
```diff
@@ -770,6 +770,126 @@ export class Gtxna extends Op {
...
+
+// branch conditionally if top of stack is non zero
+export class BranchIfNotZero extends Op {
+  readonly label: string;
+  readonly interpreter: Interpreter;
+
+  /**
+   * @param args words list extracted from line
+   * @param line line number in TEAL file
+   * @param interpreter interpreter
```
Comment (amityadav0) [inline packages/algorand-js/src/interpreter/opcode-list.ts:847]:
```suggestion
   * @param interpreter interpreter object
```

File: packages/algorand-js/src/interpreter/opcode-list.ts [hunk 7]
```diff
@@ -770,6 +770,126 @@ export class Gtxna extends Op {
...
+}
+
+// use last value on stack as success value; end
+export class Return extends Op {
+  readonly interpreter: Interpreter;
+
+  /**
+   * @param args words list extracted from line
+   * @param line line number in TEAL file
+   * @param interpreter interpreter
```
Comment (amityadav0) [inline packages/algorand-js/src/interpreter/opcode-list.ts:873]:
```suggestion
   * @param interpreter interpreter object
```

**Review by amityadav0 — CHANGES_REQUESTED** [review]
Few changes

File: packages/algorand-js/src/errors/errors-list.ts (hunk 2 above)
Comment (ratik21) [reply packages/algorand-js/src/errors/errors-list.ts:147]: yes

File: packages/algorand-js/src/interpreter/opcode-list.ts (hunk 3 above)
Comment (ratik21) [reply packages/algorand-js/src/interpreter/opcode-list.ts:784]: Thanks for catching.

File: packages/algorand-js/test/src/interpreter/opcode-list.ts [hunk 8]
```diff
@@ -1406,6 +1411,110 @@ describe("Teal Opcodes", function () {
...
 
+  describe("Branch Ops", function () {
+    const stack = new Stack<StackElem>();
+    const interpreter = new Interpreter();
+    const logic = [
+      new Int(['1'], 0),
+      new Int(['2'], 1),
+      new Branch(["dumb"], 2, interpreter),
+      new Int(['3'], 3),
+      new Label(["dumb"], 4),
```
Comment (amityadav0) [inline packages/algorand-js/test/src/interpreter/opcode-list.ts:1422]:
```suggestion
      new Label(["dumb:"], 4),
```

**Review by amityadav0 — APPROVED** [review]

**Pull request merged**